src/librbd/io/SimpleSchedulerObjectDispatch.cc

```diff
@@ -0,0 +1,421 @@
+template <typename I>
+bool SimpleSchedulerObjectDispatch<I>::flush(
+    io::FlushSource flush_source, const ZTracer::Trace &parent_trace,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  dispatch_delayed_requests(on_finish);
+
+  return false;
+}
+
+template <typename I>
+bool SimpleSchedulerObjectDispatch<I>::can_merge_write(
+    uint64_t object_no, uint64_t object_off, ceph::bufferlist&& data,
+    const ::SnapContext &snapc, int op_flags, int object_dispatch_flags,
+    Context* on_dispatched) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+
+  if (m_latency_stats && !m_latency_stats->is_ready()) {
+    ldout(cct, 20) << "latency stats not collected yet" << dendl;
+    return false;
+  }
+
+  auto it = m_delayed_requests.find(object_no);
+  if (it == m_delayed_requests.end()) {
+    ldout(cct, 20) << "no pending requests" << dendl;
+    return false;
+  }
+
+  auto &delayed_requests = it->second;
+  if (!delayed_requests.requests.empty()) {
+    if (snapc.seq != delayed_requests.snapc.seq ||
+        op_flags != delayed_requests.op_flags) {
+      return false;
+    }
+    if (object_off + data.length() == delayed_requests.offset) {
+      delayed_requests.offset = object_off;
+      data.append(std::move(delayed_requests.data));
+      delayed_requests.data.clear();
+    } else if (object_off != delayed_requests.offset +
+                             delayed_requests.data.length()) {
+      return false;
+    }
+  } else {
+    delayed_requests.dispatch_time = ceph_clock_now();
+    if (m_latency_stats) {
+      delayed_requests.dispatch_time += utime_t(0, m_latency_stats->avg() / 2);
+    } else {
+      delayed_requests.dispatch_time += utime_t(0, m_max_delay * 1000000);
+    }
+    delayed_requests.offset = object_off;
+    delayed_requests.snapc = snapc;
+    delayed_requests.op_flags = op_flags;
+
+    m_dispatch_queue.push_back(&delayed_requests);
+    if (m_dispatch_queue.front() == &delayed_requests) {
+      rescheduler_dispatch_delayed_requests();
+    }
+  }
+
+  delayed_requests.data.append(data);
+  delayed_requests.object_dispatch_flags |= object_dispatch_flags;
+  delayed_requests.requests.push_back(on_dispatched);
+
+  ldout(cct, 20) << dendl;
+
+  return true;
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  for (auto &it : m_delayed_requests) {
+    dispatch_delayed_requests(it.first, on_finish);
+  }
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    uint64_t object_no, Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+
+  dispatch_delayed_requests(object_no);
+
+  auto res = m_delayed_requests.insert({object_no, {object_no}});
```

Comment:
If there is no in-flight IO to a given object, you should release the IO immediately and not queue it for the plug period.

Reply:
It does not queue the request. It just registers it (so the scheduler knows there is a request in-flight) and wraps `on_finish` thus when this request eventually finishes it  will dispatch delay requests (if they come meantime). Or did I miss your point?

Reply:
I probably got mixed up. If it gets added to ```m_delayed_requests``` then wouldn't another write to the same object attempt to merge w/ it (even though it's not mergeable)? I guess it's just because ```requests``` would be empty that it just re-uses the existing request object?

It might be clearer to have separate collections to track in-flight IOs to objects (could be as simple as a object_no -> count of in-flight IOs) and then another collection for pending requests (for merge check) that is removed when the IO request is dispatched (or moved/added to another in-flight request collection if needed for tracking extra state).

Reply:
When the first request for an object comes it registers `ObjectDelayedRequests` instance in `m_delayed_requests`. This instance has `requests` list, which is empty initially and will contain mergable requests (when they come). It also contains `dispatch_seq` field which identifies this instance with the request that created it. When a new request comes to the object and is identified as `can_merge` it is added to the (delayed) `requests` list. If a request comes that can not be merged it dispatches the delayed requests and removes `ObjectDelayedRequests` instance. When the initial (first) request completes it checks if its `ObjectDelayedRequests` instance still exists and if it does not (there is no an instance in `m_delayed_requests` for this object or it does not match `dispatch_seq`) it just finishes otherwise it it dispatches the delayed requests and removes `ObjectDelayedRequests` instance. 

Probably `ObjectDelayedRequests` is not the best name and adds confusion.

I think I can try your approach though right now it does not look much clearer to me. Probably because I just get used to mine.

Reply:
Maybe just ```requests``` should be renamed to ```delayed_requests```? Hopefully, at the end it won't require a complete re-factor to eventually add support for out-of-order merging.